pkg/querier/tripperware/query_attribute_matcher.go

```diff
@@ -0,0 +1,237 @@
+	}
+
+	return true
+}
+
+func matchAttributeForMetadataQuery(attribute validation.QueryAttribute, r *http.Request, now time.Time) bool {
+	if err := r.ParseForm(); err != nil {
+		return false
+	}
+	if attribute.Regex != "" && attribute.Regex != ".*" && attribute.CompiledRegex != nil {
+		atLeastOneMatched := false
+		for _, matcher := range r.Form["match[]"] {
+			if attribute.CompiledRegex.MatchString(matcher) {
+				atLeastOneMatched = true
+				break
+			}
+		}
+		if !atLeastOneMatched {
+			return false
+		}
+	}
+
+	startTime, _ := util.ParseTime(r.FormValue("start"))
+	endTime, _ := util.ParseTime(r.FormValue("end"))
+
+	if !isWithinTimeAttributes(attribute.TimeWindow, now, startTime, endTime) {
+		return false
+	}
+
+	if !isWithinTimeRangeAttribute(attribute.TimeRangeLimit, startTime, endTime) {
+		return false
+	}
+
+	if attribute.UserAgentRegex != "" && attribute.UserAgentRegex != ".*" && attribute.CompiledUserAgentRegex != nil {
+		if !attribute.CompiledUserAgentRegex.MatchString(r.Header.Get("User-Agent")) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func isWithinTimeAttributes(timeWindow validation.TimeWindow, now time.Time, startTime, endTime int64) bool {
+	if timeWindow.Start == 0 && timeWindow.End == 0 {
+		return true
+	}
+
+	if timeWindow.Start != 0 {
+		startTimeThreshold := now.Add(-1 * time.Duration(timeWindow.Start).Abs()).Add(-1 * time.Minute).Truncate(time.Minute).UnixMilli()
+		if startTime == 0 || startTime < startTimeThreshold {
+			return false
+		}
+	}
+
+	if timeWindow.End != 0 {
+		endTimeThreshold := now.Add(-1 * time.Duration(timeWindow.End).Abs()).Add(1 * time.Minute).Truncate(time.Minute).UnixMilli()
+		if endTime == 0 || endTime > endTimeThreshold {
+			return false
+		}
+	}
+
+	return true
+}
+
+func isWithinTimeRangeAttribute(limit validation.TimeRangeLimit, startTime, endTime int64) bool {
+	if limit.Min == 0 && limit.Max == 0 {
+		return true
+	}
+
+	if startTime == 0 || endTime == 0 {
+		return false
+	}
+
+	timeRangeInMillis := endTime - startTime
+
+	if limit.Min != 0 && time.Duration(limit.Min).Milliseconds() > timeRangeInMillis {
+		return false
+	}
+	if limit.Max != 0 && time.Duration(limit.Max).Milliseconds() < timeRangeInMillis {
+		return false
+	}
+
+	return true
+}
+
+func isWithinQueryStepLimit(queryStepLimit validation.QueryStepLimit, r *http.Request, expr parser.Expr) bool {
+	if queryStepLimit.Min == 0 && queryStepLimit.Max == 0 {
+		return true
+	}
+	var stepLimitChecked bool
+
+	if step, err := util.ParseDurationMs(r.FormValue("step")); err == nil {
+		if queryStepLimit.Min != 0 && time.Duration(queryStepLimit.Min).Milliseconds() > step {
+			return false
+		}
+		if queryStepLimit.Max != 0 && time.Duration(queryStepLimit.Max).Milliseconds() < step {
+			return false
+		}
+		stepLimitChecked = true
+	}
+
+	subQueryStepWithinLimit := true
+	parser.Inspect(expr, func(node parser.Node, nodes []parser.Node) error {
+		e, ok := node.(*parser.SubqueryExpr)
+		if ok && e.Step != 0 {
```

Comment:
We don't handle it. I think use case for step limit going to be used to eliminate/include tiny steps. If subquery step is not provided, we just ignore it, because default evaluation step is reasonable amount. Of course we can use default subquery evaluation step, but didn't want to overcomplicate, considering small chance of usability.

Reply:
Cool. I am fine to skip that. Thanks for the explanation